Centralize addition of "always on" dispatch keys.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32788 Add XLAPreAutograd key for XLA use cases that need custom autograd.
* #32787 Stop using dispatchTypeId to do checks for tensor list unwrap.
* **#32734 Centralize addition of "always on" dispatch keys.**
* #32729 Make DispatchKeyGuards accept DispatchKey::Undefined
* #32728 Rename DispatchKey::UndefinedTensorId to Undefined

VariableTensorId is the only key with this treatment today,
but BackendSelect and CompoundOp are coming soon.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628091](https://our.internmc.facebook.com/intern/diff/D19628091)

ezyang committed:
Centralize addition of "always on" dispatch keys.

VariableTensorId is the only key with this treatment today,
but BackendSelect and CompoundOp are coming soon.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit f271eec0:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f271eec03cbd8969b2fcd4b1467b81939c68e8f1).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4365736) pytorch_linux_xenial_py2_7_9_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92351256) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92351256)) :snowflake:
<details>
<summary>
<code>Jan 29 01:27:00 RuntimeError: quantization scale should be > 0</code>
</summary>

```
Jan 29 01:27:00   File "test_quantization.py", line 730, in test_quantized_rnn 
Jan 29 01:27:00     y, (h, c) = cell_dq(x, (h, c)) 
Jan 29 01:27:00   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/nn/modules/module.py", line 540, in __call__ 
Jan 29 01:27:00     result = self.forward(*input, **kwargs) 
Jan 29 01:27:00   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 373, in forward 
Jan 29 01:27:00     return self.forward_tensor(input, hx) 
Jan 29 01:27:00   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 334, in forward_tensor 
Jan 29 01:27:00     input, hx, batch_sizes, max_batch_size, sorted_indices) 
Jan 29 01:27:00   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 315, in forward_impl 
Jan 29 01:27:00     self.batch_first, dtype=self.dtype, use_dynamic=True) 
Jan 29 01:27:00 RuntimeError: quantization scale should be > 0 
Jan 29 01:27:00  
Jan 29 01:27:00 ---------------------------------------------------------------------- 
Jan 29 01:27:00 Ran 36 tests in 57.534s 
Jan 29 01:27:00  
Jan 29 01:27:00 FAILED (errors=1, skipped=1) 
Jan 29 01:27:00  
Jan 29 01:27:00 Generating XML reports... 
Jan 29 01:27:01 Traceback (most recent call last): 
Jan 29 01:27:01   File "test/run_test.py", line 456, in <module> 
Jan 29 01:27:01     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

izdeby commented:
This PR made me think of a centralized place for fallback registrations. Do we need a mechanism like that?

ezyang commented:
> This PR made me think of a centralized place for fallback registrations as the number of DispatchKeys that are available is growing. Do we need a mechanism like that? OR its too soon to make this call as there are only 2 keys like that - variable and backendSelect?

Yeah, I dunno. It's not strictly necessary to centralize registrations; e.g., CPU/CUDA are registered in CPUType/CUDAType, and then Variable is registered in VariableType. But I agree it's not that clear what the most obvious place to put registrations is.

ezyang committed:
Update on "Centralize addition of "always on" dispatch keys."


VariableTensorId is the only key with this treatment today,
but BackendSelect and CompoundOp are coming soon.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628091](https://our.internmc.facebook.com/intern/diff/D19628091)

[ghstack-poisoned]

ezyang committed:
Update on "Centralize addition of "always on" dispatch keys."


VariableTensorId is the only key with this treatment today,
but BackendSelect and CompoundOp are coming soon.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628091](https://our.internmc.facebook.com/intern/diff/D19628091)

[ghstack-poisoned]
